sources/packages/backend/apps/api/src/services/cas-invoice-batch/cas-invoice-batch-report.service.ts

```diff
@@ -0,0 +1,191 @@
+import { Injectable } from "@nestjs/common";
+import { InjectRepository } from "@nestjs/typeorm";
+import { CASInvoice, CASInvoiceBatch } from "@sims/sims-db";
+import { Repository } from "typeorm";
+import { unparse } from "papaparse";
+import { CustomNamedError, getPSTPDTDateTime } from "@sims/utilities";
+import { CASInvoiceBatchReport } from "./cas-invoice-batch.models";
+import { CAS_INVOICE_BATCH_NOT_FOUND } from "../../constants";
+
+@Injectable()
+export class CASInvoiceBatchReportService {
+  constructor(
+    @InjectRepository(CASInvoiceBatch)
+    private readonly casInvoiceBatchRepo: Repository<CASInvoiceBatch>,
+  ) {}
+
+  /**
+   * Generates a report for a specific CAS invoice batch.
+   * Retrieves the batch data and its invoices, then creates a detailed
+   * report with information on each invoice and its related awards.
+   * This report is meant to represent a snapshot of the batch data that
+   * will be later submitted to CAS.
+   * @param invoiceBatchId ID of the CAS invoice batch to generate the report for.
+   * @returns a CSV formatted string containing the report data for the specified invoice batch.
+   */
+  async getCASInvoiceBatchesReport(
+    invoiceBatchId: number,
+  ): Promise<CASInvoiceBatchReport> {
+    const reportData = await this.getCASInvoiceBatchesReportData(
+      invoiceBatchId,
+    );
+    if (!reportData) {
+      throw new CustomNamedError(
+        `CAS invoice batch with ID ${invoiceBatchId} not found.`,
+        CAS_INVOICE_BATCH_NOT_FOUND,
+      );
+    }
+    const report = reportData.casInvoices.flatMap((invoice) => {
+      // Rows for a single invoice.
+      // One row per award will be generated.
+      const invoiceReportRows: Record<string, string>[] = [];
+      // Invoice master details. Each award will have this information.
+      const reportInvoiceMaster = this.createReportInvoiceMaster(invoice);
+      // Get all unique awards for the invoice. Each award will be a row in the report.
+      const awardCodes = new Set(
+        invoice.casInvoiceDetails.map(
+          (detail) => detail.casDistributionAccount.awardValueCode,
+        ),
+      );
+      // Get the awards distribution details to be appended to the same row.
+      awardCodes.forEach((awardCode) => {
```

Comment:
Why is `invoice.casInvoiceDetails` wasn't used to iterate for invoice entries?

Reply:
`casInvoiceDetails` has two (or as many active accounts) per award added, which means, for instance, two awards will generate four invoice details to be submitted to CAS.

Reply:
Adjusted the name to something like `downloadCASInvoiceBatchReport` or `getCASInvoiceBatchReport` for Web/API/Services.